### ISSUE TITLE
How  to open the folder containing the current file by the default explorer of the OS?

### ISSUE THREAD

**Issue by bixuanzju** [syl20bnr/spacemacs#6766]
As the title, I know `SPC f o` to open the file in the external app, often I also want to open the folder containing the current file by the default explorer (Finder for example in MacOS). Is there a command for that?

**Comment by TheBB** [comment]
No, there's no command for that. You can open it from dired using `SPC f o`.

**Comment by bixuanzju** [comment]
@TheBB Well, that would require more steps, for example, `SPC f j`, and then move the cursor to `.`, and `SPC f o`. I was thinking maybe we could enhance `SPC f o` such that if the universal argument is used, open the folder containing the current file. Actually I go about implementing this

```elisp
(defun spacemacs/open-in-external-app (arg)
  "Open current file in external application.
If the universal prefix argument is used then open the folder
containing the current file by the default explorer."
  (interactive "P")
  (let ((file-path (if (eq major-mode 'dired-mode)
                       (dired-get-file-for-visit)
                     (buffer-file-name))))
    (if file-path
        (cond
         ((spacemacs/system-is-mswindows)
          (if arg (w32-shell-execute "explore" (replace-regexp-in-string "/" "\\" default-directory t t))
            (w32-shell-execute "open" (replace-regexp-in-string "/" "\\\\" file-path))))
         ((spacemacs/system-is-mac)
          (if arg (shell-command "open .")
            (shell-command (format "open \"%s\"" file-path))))
         ((spacemacs/system-is-linux)
          (let ((process-connection-type nil))
            (if arg (start-process "" nil "xdg-open" ".")
              (start-process "" nil "xdg-open" file-path)))))
      (message "No file associated to this buffer."))))
``` 
What do you think? Worth a PR?

**Comment by bmag** [comment]
@bixuanzju It's kind of weird to use "explore" for directories when "open" should already open explorer when `file-path` points to a directory. I suggest a different approach:
- don't use `arg` in `spacemacs/open-in-external-app`
- change `spacemacs/open-in-external-app` to take an optional `file-path` argument
- when `spacemacs/open-in-external-app` is called without a `file-path` argument, use the current method to set `file-path` (I mean `(if (eq major mode 'dired-mode) ...)`).
- implement `spacemacs/open-directory-in-external-app`, which calls `(spacemacs/open-in-external-app default-directory)` to open current directory in external file browser.
- if you'd like, implement an `(interactive "P")` function that calls `spacemacs/open-in-external-app` or `spacemacs/open-directory-in-external-app` depending on the value of prefix argument.
- add/change key-bindings as appropriate

This makes it possible to use `spacemacs/open-in-external-app` in other situations that may arise in the future, and is a more general solution.

**Comment by bixuanzju** [comment]
@bmag Thanks for the suggestions! I made the following changes
+  Instead of making `spacemacs/open-in-external-app` take an optional `file-path`, it takes one actual argument, it's the job of the caller function to supply `file-path`. Also make it an internal function.
+ New interactive function `spacsmacs/open-file-or-directory-in-external-app`, depending on the value of prefix argument, it opens the current file or directory in external application.

```elisp
(defun spacemacs//open-in-external-app (file-path)
  "Open `file-path' in external application."
    (if file-path
        (cond
         ((spacemacs/system-is-mswindows) (w32-shell-execute "open" (replace-regexp-in-string "/" "\\\\" file-path)))
         ((spacemacs/system-is-mac) (shell-command (format "open \"%s\"" file-path)))
         ((spacemacs/system-is-linux) (let ((process-connection-type nil))
                                        (start-process "" nil "xdg-open" file-path))))
      (message "No file associated to this buffer.")))


(defun spacsmacs/open-file-or-directory-in-external-app (arg)
  "Open current file in external application.
If the universal prefix argument is used then open the folder
containing the current file by the default explorer."
  (interactive "P")
  (if arg (spacemacs//open-in-external-app (expand-file-name default-directory))
    (let ((file-path (if (eq major-mode 'dired-mode)
                         (dired-get-file-for-visit)
                       buffer-file-name)))
      (spacemacs//open-in-external-app file-path))))
```

**Comment by bmag** [comment]
Looks good to me, except for:
- The error message doesn't make sense in `spacemacs//open-in-external-app`, you should move it to the new function.
- The `if` in `spacsmacs/open-file-or-directory-in-external-app` is formatted weirdly. The usual way is to write:
```elisp
(if arg
    (spacemacs//open-in-external-app ...)
  (let (...)
    ...))
```

**Comment by bixuanzju** [comment]
@TheBB Thanks! See #6779

**Comment by houli** [comment]
Does `SPC b f` not do this? That's what I use.

**Comment by TheBB** [comment]
Only on OSX.

**Comment by d12frosted** [comment]
I also use `SPC p O` to open project's root folder in external app.

**Comment by bixuanzju** [comment]
@d12frosted which command exactly are you referring to? In my machine `SPC p O` isn't bound to any command, but `SPC p o` to `projectile-multi-occur`.

**Comment by d12frosted** [comment]
Oh sorry. I meant, that I defined a function to open project root and bound it to `SPC p O`. If you are interested, I can share it once I am home.

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
